stumpy/motifs.py

```diff
@@ -0,0 +1,412 @@
+
+    max_distance : float or function
+        Maximum distance between a motif candidate `Q` and a subsequence `S` and  for
+        `S` to be considered a match of the motif represented by `Q`.
+        If a function, then it has to be a function of one argument `D`, which will be
+        the distance profile of `Q` with `T` (a 1D numpy array of size `n-m+1`).
+        If None, defaults to `max(np.mean(D) - 2 * np.std(D), np.min(D))`, meaning
+        that least the closest match will be returned.
+
+    max_matches : int
+        The maximum amount of similar matches to be returned. The resulting
+        matches are sorted by distance, so a value of `10` means that the
+        indices of the most similar `10` subsequences is returned. If `None`,
+        all matches in the given tolerance range are returned.
+
+    max_motifs : int
+        Maximum number of different motifs to return.
+
+    normalize : bool
+        When set to `True`, this z-normalizes subsequences prior to computing distances.
+        Otherwise, this function gets re-routed to its complementary non-normalized
+        equivalent set in the `@core.non_normalized` function decorator.
+
+    Return
+    ------
+    top_k_indices : list
+        List of the indices of all matches of the top k motifs, sorted by distance
+        from the motif representative. The motif representative starts at the first
+        index and is the subsequence with the lowest matrix profile value that is part
+        of a motif.
+
+    top_k_values : ndarray
+        List of the distances of all matches of the top k motifs to the
+        motif representative.
+    """
+    n = T.shape[1]
+    l = P.shape[1]
+    m = n - l + 1
+
+    top_k_indices = []
+    top_k_values = []
+
+    candidate_idx = np.argmin(P[-1])
+    while len(top_k_indices) < max_motifs:
+        profile_value = P[-1, candidate_idx]
+        if np.isinf(profile_value):  # pragma: no cover
+            break
+
+        Q = T[:, candidate_idx : candidate_idx + m]
+
+        query_matches = match(
+            Q,
+            T,
+            M_T=M_T,
+            Σ_T=Σ_T,
+            excl_zone=excl_zone,
+            max_matches=None,
+            max_distance=max_distance,
+            normalize=normalize,
+        )
+
+        if len(query_matches) > min_neighbors:
+            top_k_indices.append(query_matches[:max_matches, 0])
+            top_k_values.append(query_matches[:max_matches, 1])
+
+        if len(query_matches) > 0:
+            for idx in query_matches[:, 0]:
+                core.apply_exclusion_zone(P, idx, excl_zone)
+        else:  # pragma: no cover
+            # I no match was found, still apply an exclusion zone around the candidate
+            # index, we do not want to consider it as a motif candidate again.
+            core.apply_exclusion_zone(P, candidate_idx, excl_zone)
+
+        candidate_idx = np.argmin(P[-1])
+
+    return top_k_indices, top_k_values
+
+
+def motifs(
```

Comment:
So, there is a fundamental question of whether or not we should add the query motif into the list-of-lists?  Currently, we identify a query motif but it looks like we don't add that query subsequence to the results. This means that it is up to the user to figure out what the query motif is. 

I think we should include the query motif (and index) as the first element of each list.

Reply:
Good points! Thank you